contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,821 @@
+  continueStatusInternal(status);
+}
+
+void Filter::continueStatusInternal(GolangStatus status) {
+  ProcessorState& state = getProcessorState();
+  ASSERT(state.isThreadSafe());
+  auto saved_state = state.state();
+
+  if (local_reply_waiting_go_) {
+    ENVOY_LOG(warn,
+              "other filter already trigger sendLocalReply, ignoring the continue status: {}, "
+              "state: {}, phase: {}",
+              int(status), state.stateStr(), state.phaseStr());
+
+    continueEncodeLocalReply(state);
+    return;
+  }
+
+  auto done = state.handleGolangStatus(status);
+  if (done) {
+    switch (saved_state) {
+    case FilterState::ProcessingHeader:
+      // NP: should process data first filter seen the stream is end but go doesn't,
+      // otherwise, the next filter will continue with end_stream = true.
+
+      // NP: it is safe to continueDoData after continueProcessing
+      // that means injectDecodedDataToFilterChain after continueDecoding while stream is not end
+      if (state.isProcessingEndStream() || !state.isStreamEnd()) {
+        state.continueProcessing();
+      }
+      break;
+
+    case FilterState::ProcessingData:
+      state.continueDoData();
+      break;
+
+    case FilterState::ProcessingTrailer:
+      state.continueDoData();
+      state.continueProcessing();
+      break;
+
+    default:
+      ASSERT(0, "unexpected state");
+    }
+  }
+
+  // TODO: state should also grow in this case
+  // state == WaitingData && bufferData is empty && seen trailers
+
+  auto current_state = state.state();
+  if ((current_state == FilterState::WaitingData &&
+       (!state.isBufferDataEmpty() || state.getEndStream())) ||
+      (current_state == FilterState::WaitingAllData && state.isStreamEnd())) {
+    auto done = doDataGo(state, state.getBufferData(), state.getEndStream());
+    if (done) {
+      state.continueDoData();
+    } else {
+      // do not process trailers when data is not finished
+      return;
+    }
+  }
+
+  Thread::ReleasableLockGuard lock(mutex_);
+  if (state.state() == FilterState::WaitingTrailer && trailers_ != nullptr) {
+    auto trailers = trailers_;
+    lock.release();
+    auto done = doTrailerGo(state, *trailers);
+    if (done) {
+      state.continueProcessing();
+    }
+  }
+}
+
+void Filter::sendLocalReplyInternal(
+    Http::Code response_code, absl::string_view body_text,
+    std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+    Grpc::Status::GrpcStatus grpc_status, absl::string_view details) {
+  ENVOY_LOG(debug, "sendLocalReply Internal, response code: {}", int(response_code));
+
+  ProcessorState& state = getProcessorState();
+
+  if (local_reply_waiting_go_) {
+    ENVOY_LOG(warn,
+              "other filter already invoked sendLocalReply or encodeHeaders, ignoring the local "
+              "reply from go, code: {}, body: {}, details: {}",
+              int(response_code), body_text, details);
```

Comment:
Same question. Audit all warns in the filter code.

Reply:
Done.